drivers/misc/cldatadrv/rcar_cluster_drv.c

```diff
@@ -0,0 +1,330 @@
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/wait.h>
+#include <linux/kthread.h>
+#include <linux/dma-mapping.h>
+#include <linux/miscdevice.h>
+
+#include <linux/rpmsg.h>
+#include <linux/of_reserved_mem.h>
+#include <linux/atomic.h>
+#include <linux/skbuff.h>
+#include <uapi/linux/rpmsg.h>
+#include "linux/cdev.h"
+#include "r_taurus_cluster_protocol.h"
+#include "rcar_cluster_drv.h"
+
+static DEFINE_IDR(misc_idr);
+static DEFINE_MUTEX(idr_lock);
+
+#define MAX_MISC_RPMSG	1
+
+#define RPMSG_DEV_MAX	(MINORMASK + 1)
+
+#define RCAR_CLUSTER_NAME     "rcar-cluster-drv"
+
+#define RCAR_IO_SPEED  1
+#define RCAR_IO_GEAR   2
+
+#define dev_to_clusterdev(dev) container_of(dev, rcar_cluster_device_t, dev)
+
+static int rpmsg_cluster_probe(struct rpmsg_device *rpdev);
+static int rpmsg_cluster_cb(struct rpmsg_device* rpdev, void* data, int len,
+			void* priv, u32 src);
+static void rpmsg_cluster_remove(struct rpmsg_device *rpdev);
+static ssize_t misc_write(struct file *p_file,
+		     const char __user *buf,
+		     size_t count,
+		     loff_t *f_pos);
+static int misc_open(struct inode *inode, struct file *p_file);
+static int misc_release(struct inode *inode, struct file *p_file);
+
+static struct rpmsg_device_id rpmsg_driver_cluster_id_table[] = {
+	{ .name	= "taurus-cluster" },
+	{ },
+};
+
+static struct rpmsg_driver rpmsg_cluster_drv = {
+	.drv.name	= KBUILD_MODNAME,/*"rpmsg_chrdev",*/
+	.drv.probe_type = PROBE_PREFER_ASYNCHRONOUS,
+	.id_table	= rpmsg_driver_cluster_id_table,
+	.probe		= rpmsg_cluster_probe,
+	.callback	= rpmsg_cluster_cb,
+	.remove		= rpmsg_cluster_remove,
+};
+
+static const struct file_operations misc_fops = {
+	.owner = THIS_MODULE,
+	.open = misc_open,
+	.write = misc_write,
+	.release = misc_release,
+};
+
+static atomic_t rpmsg_id_counter = ATOMIC_INIT(0);
+
+static int cluster_taurus_get_uniq_id(void) {
+	return atomic_inc_return(&rpmsg_id_counter);
+}
+
+static int send_msg(struct rpmsg_device *rpdev, 
+					taurus_cluster_data_t *data,
+					taurus_cluster_res_msg_t* res_msg) {
+	int ret = 0;
+	R_TAURUS_CmdMsg_t msg;
+	struct taurus_event_list* event = NULL;
+	rcar_cluster_device_t *clusterdrv = NULL;
+
+	event = devm_kzalloc(&rpdev->dev, sizeof(*event), GFP_KERNEL);
+	if (!event) {
+		dev_err(&rpdev->dev, "send_msg: Can't allocate memory for taurus event\n");
+		return -ENOMEM;
+	}
+	event->result = devm_kzalloc(&rpdev->dev, sizeof(*event->result), GFP_KERNEL);
+    
+	if (!event->result) {
+       	dev_err(&rpdev->dev, "send_msg: Can't allocate memory for taurus event->result\n");
+		devm_kfree(&rpdev->dev, event);
+		return -ENOMEM;
+ 	}
+
+	clusterdrv = (rcar_cluster_device_t*)dev_get_drvdata(&rpdev->dev);
+
+	if(!clusterdrv) {
+		dev_err(&rpdev->dev, "send_msg: Can't get data type rcar_cluster_device*\n");
+		devm_kfree(&rpdev->dev, event->result);
+		devm_kfree(&rpdev->dev, event);		
+		return -ENOMEM;
+	}
+
+	msg.Id = cluster_taurus_get_uniq_id();
+	msg.Channel = 0x80;
```

Comment:
Why `0x80`? Looks like magic number

Reply:
No, this is the channel number has been defined in cr7 application to identify the source of the message